src/widget/wlibrarytableview.cpp

```diff
@@ -53,80 +61,128 @@ WLibraryTableView::WLibraryTableView(QWidget* parent,
-}
 
 void WLibraryTableView::moveSelection(int delta) {
     QAbstractItemModel* pModel = model();
 
     if (pModel == nullptr) {
         return;
     }
-
-    while(delta != 0) {
-        // TODO(rryan) what happens if there is nothing selected?
-        QModelIndex current = currentIndex();
-        if(delta > 0) {
-            // i is positive, so we want to move the highlight down
-            int row = current.row();
-            if (row + 1 < pModel->rowCount()) {
-                selectRow(row + 1);
+    QItemSelectionModel* currentSelection = selectionModel();
+    clearSelection();
+    currentSelection->clearSelection();
+    QModelIndex newIndex;
+    while (delta != 0) {
+        QModelIndex currentIndex = currentSelection->currentIndex();
+        if (currentIndex.isValid()) {
+            int row = currentIndex.row();
+            if (delta > 0) {
+                // i is positive, so we want to move the highlight down
+                if (row + 1 < pModel->rowCount()) {
+                    newIndex = currentIndex.sibling(row + 1, 0);
+                } else {
+                    // we wrap around at the end of the list so it is faster to get
+                    // to the top of the list again
+                    newIndex = currentIndex.sibling(0, 0);
+                }
+                delta--;
+            } else {
+                if (row - 1 >= 0) {
+                    newIndex = currentIndex.sibling(row - 1, 0);
+                } else {
+                    newIndex = pModel->index(pModel->rowCount() - 1, 0);
+                }
+                delta++;
             }
-
-            delta--;
         } else {
-            // i is negative, so we want to move the highlight up
-            int row = current.row();
-            if (row - 1 >= 0) {
-                selectRow(row - 1);
+            // no selection, so select the first or last element depending on delta
+            if (delta > 0) {
+                newIndex = pModel->index(0, 0);
+                delta--;
+            } else {
+                newIndex = pModel->index(pModel->rowCount() - 1, 0);
+                delta++;
             }
-
-            delta++;
         }
+        // this wired combination works when switching between models
+        setCurrentIndex(newIndex);
+        currentSelection->select(newIndex, QItemSelectionModel::ClearAndSelect);
+        // why does scrollTo not work ?
+        // scrollTo(newIndex);
+        selectRow(newIndex.row());
     }
 }
 
-void WLibraryTableView::saveVScrollBarPos(TrackModel* key){
-    m_vScrollBarPosValues[key] = verticalScrollBar()->value();
+void WLibraryTableView::saveTrackModelState(const QAbstractItemModel* model, const QString& key) {
+    VERIFY_OR_DEBUG_ASSERT(model) {
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(!key.isEmpty()) {
+        return;
+    }
+    ModelState* state;
+    if (m_vModelState.contains(key)) {
+        state = m_vModelState[key];
```

Comment:
This is a double lookup. Please use m_vModelState[key] directly and check the result for null.